### PR TITLE
docs: new API guide

### PR DESCRIPTION
**Pull request by mbshields**
**What type of PR is this?**

documentation

**Which issue does this PR fix**:

[Feat]: Document zot endpoints in the developer guide #115

**What does this PR do / Why do we need it**:

documents API

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/developer-guide/api-guide.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+| /auth/apikey | With query parameters, creates, lists, or deletes API keys | Available when API key authentication is enabled in the configuration file ("apikey": true). |
+| /auth/callback | Redirects a login to a provider service | Available when OpenID authentication is enabled in the configuration file. |
+| /oras/artifacts/v1/ | ORAS endpoints | Always enabled. |
+| /metrics | Displays extended metrics | Available when the `metrics` build label is used and the `metrics` extension is enabled  in the configuration file. |
+| /swagger/v2/ | Displays the OpenAPI (formerly Swagger) documentation  | Enabled by using the `debug` build label (`make binary-debug`).
+| /v2/ | [OCI specification endpoints](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#endpoints) | Always available. |
+| /v2/_oci/ext/discover | Discover extensions per the OCI specification | Always available. |
+| /v2/metrics | Displays basic metrics | Available when the `metrics` extension is disabled, whether or not the `metrics` build label was used. |
```
Comment (andaaron) [inline docs/developer-guide/api-guide.md:166]: Note @adodon2go is reviewing this endpoint. There may be changes.

File: docs/developer-guide/api-guide.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+| /v2/_zot/ext/mgmt | Mgmt extension endpoints | Enabled by using the `mgmt` build label and enabling both the `search` and `ui` extensions in the configuration file. |
```
Comment (andaaron) [inline docs/developer-guide/api-guide.md:167]:
```suggestion
| /v2/_zot/ext/mgmt | Mgmt extension endpoints | Enabled by using the `mgmt` build label and enabling the `search` extension in the configuration file. |
```

File: docs/developer-guide/api-guide.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+| /v2/_zot/ext/cosign | Uploads keys for signature verification | Enabled by using the `imagetrust` build label and enabling the `trust` extension with the `cosign` option enabled. |
+| /v2/_zot/ext/notation | With query parameters, uploads certificates for signature verification | Enabled by using the `imagetrust` build label and enabling the `trust` extension with the `notation` option enabled. |
+| /v2/_zot/ext/search | Enhanced search | Enabled by using the `search` build label and enabling the `search` extension in the configuration file. |
+| /v2/_zot/ext/userprefs | User preferences endpoints | Enabled by using the `userprefs` build label and enabling both the `search` and the `ui` extensions in the configuration file. |
+| /v2/_zot/debug/graphql-playground | [GraphQL](https://graphql.org/) playground | Enabled by using the `debug` build label. |
```
Comment (andaaron) [inline docs/developer-guide/api-guide.md:172]: This one could be more advertised.

File: docs/developer-guide/api-guide.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+!!! note "Mike's Questions" 
+
+    Does the `ui` extension in the config file enable/disable the API? What is '/' in the API?
+
+    Do we need to document the URL query parameters and API payloads here or should we link to the relevant articles for expanded details?  
+    
+    Should we at least indicate in the API table that parameters or payloads are supported for each command?
+
+    Is the API syntax finalized for uploading keys and certificates for cosign and notation?
```
[the remark this answers is not in the document]
Comment (andaaron) [reply docs/developer-guide/api-guide.md:13]: @mbshields this has been merged. Please see the updated pkg/extensions/README_imagetrust.md in https://github.com/project-zot/zot/pull/1744

File: docs/developer-guide/api-guide.md (hunk 3 above)
Comment (mbshields) [reply docs/developer-guide/api-guide.md]: Does it need an article?

File: docs/developer-guide/api-guide.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
+# Using the zot API
+
+> :point_right: The zot API implements the [OCI Distribution endpoints](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#endpoints) along with endpoints for the supported extensions.
+
+!!! note "Mike's Questions" 
+
+    Does the `ui` extension in the config file enable/disable the API? What is '/' in the API?
+
+    Do we need to document the URL query parameters and API payloads here or should we link to the relevant articles for expanded details?
```
[the remark this answers is not in the document]
Comment (mbshields) [reply docs/developer-guide/api-guide.md]: Do we need an article about it?

File: docs/developer-guide/api-guide.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+> :point_right: The zot API implements the [OCI Distribution endpoints](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#endpoints) along with endpoints for the supported extensions.
+
+!!! note "Mike's Questions" 
+
+    Does the `ui` extension in the config file enable/disable the API? What is '/' in the API?
+
+    Do we need to document the URL query parameters and API payloads here or should we link to the relevant articles for expanded details?  
+    
+    Should we at least indicate in the API table that parameters or payloads are supported for each command?
```
[the remark this answers is not in the document]
Comment (mbshields) [reply docs/developer-guide/api-guide.md]: I've converted it into the API Reference document.

File: docs/developer-guide/api-guide.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
+# Using the zot API
+
+> :point_right: The zot API implements the [OCI Distribution endpoints](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#endpoints) along with endpoints for the supported extensions.
+
+!!! note "Mike's Questions" 
+
+    Does the `ui` extension in the config file enable/disable the API? What is '/' in the API?
```
[the remark this answers is not in the document]
Comment (mbshields) [reply docs/developer-guide/api-guide.md]: OK

File: docs/developer-guide/api-guide.md (hunk 2 above)
Comment (mbshields) [reply docs/developer-guide/api-guide.md]: Fixed

File: docs/developer-guide/api-user-guide.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+
+    !!! REVIEW QUESTIONS:
+
+    1. Are these deprecated?
+        /auth/login
+        /auth/callback
+        /metrics
+        /swagger/v2/
+        /v2/metrics
```
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:48]: This is the only one which we removed.

There have been some discussions about modifying the others, but we need to clarify the proposal first.

File: docs/developer-guide/api-guide.md (hunk 1 above)
Comment (andaaron) [reply docs/developer-guide/api-guide.md]: This has been removed.

File: docs/developer-guide/api-guide.md (hunk 3 above)
Comment (andaaron) [reply docs/developer-guide/api-guide.md]: Yes, IMHO.

File: docs/developer-guide/api-user-guide.md [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+| /v2/_catalog| GET |  |
+| /v2/_oci/ext/discover| GET | Discover extensions per the OCI specification | Always available. |
+| /v2/_zot/ext/cosign| POST | Uploads keys for signature verification | Enabled by using the `imagetrust` build label and enabling the `trust` extension with the `cosign` option enabled. |
+| /v2/_zot/ext/mgmt| GET | Mgmt extension endpoints | Enabled by using the `mgmt` build label and enabling the `search` extension in the configuration file. |
+| /v2/_zot/ext/notation| POST | With query parameters, uploads certificates for signature verification | Enabled by using the `imagetrust` build label and enabling the `trust` extension with the `notation` option enabled. |
+| /v2/_zot/ext/search |  | Enhanced search | Enabled by using the `search` build label and enabling the `search` extension in the configuration file. |
+| /v2/_zot/ext/userprefs| PUT | User preferences endpoints | Enabled by using the `userprefs` build label and enabling both the `search` and the `ui` extensions in the configuration file. |
+| /v2/_zot/blobs/{digest}| DELETE, GET, HEAD |  |
```
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:34]: These urls are not correct (lines 34-38)

Comment (andaaron) [reply docs/developer-guide/api-user-guide.md:34]: Ram's suggestion for escaping `/v2/manifests/\<\>/repo`

Comment (andaaron) [reply docs/developer-guide/api-user-guide.md:34]: These are the urls defined in the spec at : https://github.com/opencontainers/distribution-spec/blob/main/spec.md#endpoints

Comment (mbshields) [reply docs/developer-guide/api-user-guide.md]: Revised OCI endpoints section

File: docs/developer-guide/api-reference.md [hunk 10]
```diff
@@ -0,0 +1,1096 @@
...
+
+### GET `/v2/{name}/tags/list`
+
+List all image tags in a repository
+
+*Parameters*
+
+|Name|In|Type|Required|Description|
+|---|---|---|---|---|
+|name|path|string|true|test|
```
Comment (andaaron) [inline docs/developer-guide/api-reference.md:1071]:
```suggestion
|name|path|string|true|repository name|
```

File: docs/developer-guide/api-user-guide.md
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:1]: I am conflicted with regards to this file name and section titles.

1. On one hand we mentioned "API" everywhere.
2. On the other hand we wanted to mention all endpoints, not just APIs (the Web-based UI for example, the Swagger UI)

We need 2 in the docs somewhere: to mention how to enable the endpoints, not just for API features, but also how to build/enable the UI endpoint, how to build/enable the Swagger endpoint.

Maybe also include the prerequisites to use the pprof and graphql playground endpoints in this guide. I don't know if they are missing by design, or there was not enough information about them available.

The information mentioned in the comment https://github.com/project-zot/project-zot.github.io/issues/115#issue-1838951730 we wanted to add was not necessarily for an "API user", but for a new zot developer working on zot to know how to troubleshoot APIs not being available. 

So are adding a separate file for zot developers on how to build/enable the endpoints? The wording in this file (and including the apikey authentication instructions) is more geared towards developers consuming the zot APIs in a client application.

File: docs/developer-guide/api-guide.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (andaaron) [reply docs/developer-guide/api-guide.md]: https://github.com/project-zot/project-zot.github.io/issues/130

**Comment by andaaron** [comment]
https://github.com/project-zot/zot/issues/1883 has been merged, some of the URLs need to be updated.

File: docs/developer-guide/api-user-guide.md [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+| /v2/{repo}/referrers/{digest}| GET | Lists referrers given a digest | 
+| /v2/{repo}/tags/list | GET | List all image tags in a repository | 
+
+### zot endpoints
+
+| Endpoint | Actions | Description | Availability |
+| -------- | ------- | ----------- | -------------|
+| / | (browser) | Web interface | Enabled by using the `ui` build label and enabling the `ui` extension in the configuration file. |
+| /auth/apikey| DELETE, GET, POST | Creates, lists, or deletes API keys | Available when API key authentication is enabled in the configuration file (`"apikey": true`). |
+| /auth/logout| POST  | Ends an API session | Available when authentication is available. This includes not only OpenID, but all session-based authentication.
```
[the remark this answers is not in the document]
Comment (mbshields) [reply docs/developer-guide/api-user-guide.md:40]: I just merged 'zot endpoints' and 'zot internal endpoints'.  Is that OK?

File: docs/developer-guide/api-user-guide.md [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+| /v2/_catalog| GET | Lists repositories in a registry. |
+| /v2/_oci/ext/discover| GET | Discover extensions per the OCI specification | 
+| /v2/{repo}/blobs/{digest}| DELETE, GET, HEAD | Lists or deletes an image's blob/layer given a digest | 
+| /v2/{repo}/blobs/uploads| POST | Creates an image blob/layer upload | 
+| /v2/{repo}/blobs/uploads/{session_id}| DELETE, GET, PATCH, PUT | Creates, lists, or deletes an image's blob/layer upload given a session_id | 
+| /v2/{repo}/manifests/{reference}| DELETE, GET, HEAD, PUT | Creates, lists, or deletes references for an image | 
+| /v2/{repo}/referrers/{digest}| GET | Lists referrers given a digest | 
+| /v2/{repo}/tags/list | GET | List all image tags in a repository | 
+
+### zot endpoints
```
[the remark this answers is not in the document]
Comment (mbshields) [reply docs/developer-guide/api-user-guide.md:34]: Moved

File: docs/developer-guide/api-user-guide.md [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+
+| Endpoint | Actions | Description | Availability |
+| -------- | ------- | ----------- | -------------|
+| / | (browser) | Web interface | Enabled by using the `ui` build label and enabling the `ui` extension in the configuration file. |
+| /auth/apikey| DELETE, GET, POST | Creates, lists, or deletes API keys | Available when API key authentication is enabled in the configuration file (`"apikey": true`). |
+| /auth/logout| POST  | Ends an API session | Available when authentication is available. This includes not only OpenID, but all session-based authentication. 
+
+### zot OCI extension endpoints
```
[the remark this answers is not in the document]
Comment (mbshields) [reply docs/developer-guide/api-user-guide.md:35]: pprof endpoint added

File: docs/developer-guide/api-user-guide.md
Comment (mbshields) [reply docs/developer-guide/api-user-guide.md:1]: I'm unclear on the distinction between API and endpoint.  I've been using them interchangeably.

Should we make a separate API guide/troubleshooting/reference doc for the developer (or admin?) and then have a simpler API guide for users?  Or is an API guide for users not necessary?

Re: Build/enable UI -- our "Use the Web Interface" article says the registry must be "built with the optional zui package".  Is this still true?  By the way, I removed the '/' endpoint from this user guide document, but the admin config guide discusses the need to enable the ui extension.

Comment (andaaron) [reply docs/developer-guide/api-user-guide.md:1]: An API is supposed to be used programmatically (by clients implemented by developers), as opposed to a user interface which is meant to be used by a person.
For example the REST API, the GraphQL API, and the UI all have endpoints, even if the UI is not an API.
Most of the endpoints we have are of interest to developers, not to end users.
The end users would typically use the zli (command line interface client) or the zui (web based client), or other tools such as skopeo.

The `/` endpoint is used by the UI, and `"Use the Web Interface" article says the registry must be "built with the optional zui package". Is this still true?` is still true.

File: docs/developer-guide/api-user-guide.md [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+| -------- | ------- | ----------- | -------------|
+| /v2/_zot/pprof/ | GET | Returns a current HTML-format profile list along with a count of currently available records for each profile. See [Performance Profiling in zot](../articles/pprof.md) for usage details. | Always enabled. |
+| /v2/_zot/debug/graphql-playground# |  | See [Using GraphQL](../articles/graphql.md) for details. | Enabled only in a `binary-debug` zot build or when the zot registry has been built with the `debug` extension label. |
+
+
+### ORAS endpoints
+
+| Endpoint | Actions | Description | Availability |
+| -------- | ------- | ----------- | -------------|
+| /oras/artifacts/v1/_zot/ \ manifests/{digest}/referrers | GET | [OCI Registry As Storage (ORAS)](https://oras.land/) endpoints | Always enabled. |
```
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:68]: This is supposed to be `/oras/artifacts/v1/{repo}/manifests/{digest}/referrers`

Comment (mbshields) [reply docs/developer-guide/api-user-guide.md:68]: Fixed

File: docs/developer-guide/api-user-guide.md
Comment (mbshields) [reply docs/developer-guide/api-user-guide.md:1]: I've added the endpoints, enabling info, and article links for pprof and graphql playground.  I'm not clear on what other changes should be made - we can have a call if needed.

File: docs/developer-guide/api-user-guide.md [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+| /v2/_zot/ext/search |  | Enhanced search | Enabled by using the `search` build label and enabling the `search` extension in the configuration file. |
+| /v2/_zot/ext/cosign| POST | Uploads keys for signature verification | Enabled by using the `imagetrust` build label and enabling the `trust` extension with the `cosign` option enabled. |
+| /v2/_zot/ext/userprefs| PUT | User preferences endpoints | Enabled by using the `userprefs` build label and enabling both the `search` and the `ui` extensions in the configuration file. |
+
+
+### zot auth endpoints
+
+| Endpoint | Actions | Description | Availability |
+| -------- | ------- | ----------- | -------------|
+| /auth/apikey| DELETE, GET, POST | Creates, lists, or deletes API keys | Available when API key authentication is enabled in the configuration file (`"apikey": true`). |
```
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:50]:
```suggestion
| /zot/auth/apikey| DELETE, GET, POST | Creates, lists, or deletes API keys | Available when API key authentication is enabled in the configuration file (`"apikey": true`). |
```

All /auth/* endpoints have been updated in https://github.com/project-zot/zot/pull/1947/files#diff-49aaa2819e35a856818ecec8c9fa7e1c79ad028d3f44bd749736353cfb51bac9 to /zot/auth/*

File: docs/developer-guide/api-user-guide.md [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+      "apikey": true
+    }
+  }
+```
+
+#### Creating your API key
+
+Before you can create or revoke an API key, you must first log in using a different authentication mechanism, such as logging in through the zot GUI. When you are logged in, you can create an API key for your identity using the following API command:
+
+    POST /auth/apikey
```
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:116]:
```suggestion
    POST /zot/auth/apikey
```

File: docs/developer-guide/api-user-guide.md [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+
+_cURL command example:_
+
+    curl -u user:password -X POST http://localhost:8080/auth/apikey -d '{"label": "myAPIKEY", "scopes": ["repo1", "repo2"], "expirationDate": "2023-08-28T17:10:05+03:00"}'
```
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:120]:
```suggestion
    curl -u user:password -X POST http://localhost:8080/zot/auth/apikey -d '{"label": "myAPIKEY", "scopes": ["repo1", "repo2"], "expirationDate": "2023-08-28T17:10:05+03:00"}'
```

File: docs/developer-guide/api-user-guide.md [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+The API key replaces a password in the API command, as shown in the following cURL example:
+
+    curl -u user:zak_e77bcb9e9f634f1581756abbf9ecd269 http://localhost:8080/v2/_catalog
+
+
+#### Removing your API key
+
+When logged in, you can revoke your own API key with the following API command:
+
+    DELETE /auth/apikey?id=$uuid
```
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:158]:
```suggestion
    DELETE /zot/auth/apikey?id=$uuid
```

File: docs/developer-guide/api-user-guide.md [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+
+_cURL command example:_
+
+    curl -u user:password -X DELETE http://localhost:8080/v2/auth/apikey?id=46a45ce7-5d92-498a-a9cb-9654b1da3da1
```
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:162]:
```suggestion
    curl -u user:password -X DELETE http://localhost:8080/zot/auth/apikey?id=46a45ce7-5d92-498a-a9cb-9654b1da3da1
```

File: docs/developer-guide/api-user-guide.md [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+
+
+#### Listing your current API keys
+
+When logged in, you can display a list of your API keys with the following API command:
+
+    GET /auth/apikey
```
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:169]:
```suggestion
    GET /zot/auth/apikey
```

File: docs/developer-guide/api-user-guide.md [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+
+_cURL command example:_
+
+    curl -u user:password -X GET http://localhost:8080/auth/apikey
```
Comment (andaaron) [inline docs/developer-guide/api-user-guide.md:173]:
```suggestion
    curl -u user:password -X GET http://localhost:8080/zot/auth/apikey 
```

File: docs/developer-guide/api-user-guide.md
Comment (andaaron) [reply docs/developer-guide/api-user-guide.md:1]: I think we're ok. I left comments for some URLs that escaped my review last week.

File: docs/developer-guide/api-reference.md [hunk 22]
```diff
@@ -0,0 +1,1096 @@
+# zot API Command Reference
+
+> :point_right: This article describes the zot REST API commands, parameters, and responses.
+
+The information presented here is adapted from the interactive OpenAPI (formerly swagger) [JSON file](https://github.com/project-zot/zot/blob/main/swagger/swagger.json) in the zot Github project.
+
+For instructions and examples of how to use the zot API, see [Using the zot API](../developer-guide/api-user-guide.md).
+
+
+## /auth/apikey
```
Comment (andaaron) [inline docs/developer-guide/api-reference.md:10]:
```suggestion
## /zot/auth/apikey
```

Please update the file everywhere we mention an /auth/ URL.

Comment (mbshields) [reply docs/developer-guide/api-reference.md]: Done

**Review by andaaron — APPROVED** [review]
LGTM.

Please solved the conflict.

**Review by andaaron — APPROVED** [review]

**Pull request merged**